tools/test/toolchains/test_toolchains.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Test the arm toolchain."""
+import pytest
+import os
+import sys
+from unittest import TestCase
+from collections import OrderedDict
+
+import mock
+
+
+ROOT = os.path.abspath(
+    os.path.join(os.path.dirname(__file__), "..", "..", "..")
+)
+sys.path.insert(0, ROOT)
+
+from tools.toolchains.arm import ARM_STD, ARM_MICRO, ARMC6
+from tools.toolchains.gcc import GCC_ARM
+from tools.toolchains.iar import IAR
+from tools.utils import NotSupportedException
+
+class TestArmToolchain(TestCase):
+    """Test Arm classes."""
+
+    def test_arm_default_lib(self):
+        """Test that linker flags are correctly added to an instance of ARM."""
+        mock_target = mock.MagicMock()
+        mock_target.core = "Cortex-M4"
+        mock_target.__dict__["supported_c_lib"] = OrderedDict([(u'arm_c_lib', [u'small'])])
+        mock_target.default_lib = "small"
+        mock_target.default_toolchain = "ARM"
+        mock_target.supported_toolchains = ["ARM", "uARM", "ARMC5", "ARMC6"]
+        arm_std_obj = ARM_STD(mock_target)
+       	arm_micro_obj = ARM_MICRO(mock_target)
+
+        mock_target.default_toolchain = "ARMC6"
+        arm_c6_obj = ARMC6(mock_target)
+
+        self.assertIn("-D__MICROLIB", arm_std_obj.flags["common"])
+        self.assertIn("-D__MICROLIB", arm_micro_obj.flags["common"])
+        self.assertIn("-D__MICROLIB", arm_c6_obj.flags["common"])
+
+        self.assertIn("--library_type=microlib", arm_std_obj.flags["ld"])
+        self.assertIn("--library_type=microlib", arm_micro_obj.flags["ld"])
+        self.assertIn("--library_type=microlib", arm_c6_obj.flags["ld"])
+
+        self.assertIn("-Wl,--library_type=microlib", arm_c6_obj.flags["cxx"])
+        self.assertIn("--library_type=microlib", arm_c6_obj.flags["asm"])
+
+    def test_arm_default_lib_exception(self):
+        """Test that exception raised when default_lib is small but supported_c_lib parameter arm_c_lib is not suppoted small lib."""
+        mock_target = mock.MagicMock()
+        mock_target.core = "Cortex-M4"
+        mock_target.microlib_supported = False
+        mock_target.default_lib = "small"
+        mock_target.__dict__["supported_c_lib"] = OrderedDict([(u'arm_c_lib', [u'std'])])
+        mock_target.default_toolchain = "ARM"
+        mock_target.supported_toolchains = ["ARM", "uARM", "ARMC5"]
+        with pytest.raises(NotSupportedException) as excinfo:
+            ARM_STD(mock_target)
+        excinfo.match(r'target.default_lib is small not supported for this target.')
```

Comment:
You can have the string `target.default_lib is small not supported for this target.` defined in a variable above since it is used multiple times in this file.
Would it be possible to use `self.assertIn(UNSUPPORTED_OPTION_EXECPTION_STRING, str(excinfo.value))`?

Reply:
Yes I can put "target.default_lib is small not supported for this target" in a variable. self.assertIn(UNSUPPORTED_OPTION_EXECPTION_STRING, str(excinfo.value)) was not working

Reply:
I tried `self.assertIn('target.default_lib is small not supported for this target.', str(excinfo.value))` and it worked. Please update it.

Reply:
In my view  this is right way to test exceptions  ```with pytest.raises(NotSupportedException) as excinfo: 
            ARM_STD(mock_target) 
       excinfo.match(r'target.default_lib is small not supported for this target.``` and it works.

Reply:
First you said it did not work and now you say in your view. It is not about personal preference or views. A lot of things work even just `assert`. However `unittest.TestCase.assertIn` is what is intended to be be used for tests.

Reply:
Actually you can even use `unittest.TestCase.assertRaisesRegexp`

Reply:
It was not working when I tried yesterday night and I preferred going with proper exception testing mechanism which gives better readability in the test suite. So I can't agree to use assert

Reply:
Again you are talking about preference whilst I am talking about what should be used as Python documentation for testing.
`unittest.TestCase.assertRaisesRegexp` should be used.

Reply:
Python documentation for testing suggested both ways we can test exception so I used pytest.raises.

Reply:
Look where my comment is! I am talking line 65. Not the rest.

Reply:
sorry i confused with pytest.raises ,i will read about "unittest.TestCase.assertRaisesRegexp "